compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -566,9 +566,30 @@ bool compareIntrinsics(const GPUMatmulShapeType &problem,
     return lhsArea > rhsArea;
   }
 
-  // Finally if everything else is the same, prefer large K size.
-  return ShapedType::getNumElements(lhs.kSizes) >
-         ShapedType::getNumElements(rhs.kSizes);
+  // Prefer large K size here.
+  int64_t lhsKSize = ShapedType::getNumElements(lhs.kSizes);
+  int64_t rhsKSize = ShapedType::getNumElements(rhs.kSizes);
+  if (lhsKSize != rhsKSize) {
+    return lhsKSize > rhsKSize;
+  }
+
+  const GPUIntrinsicType *lhsIntrinsic =
+      static_cast<const GPUIntrinsicType *>(&lhs);
+  const GPUIntrinsicType *rhsIntrinsic =
+      static_cast<const GPUIntrinsicType *>(&rhs);
```

Comment:
See https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
```suggestion
  const auto *lhsIntrinsic =
      static_cast<const GPUIntrinsicType *>(&lhs);
  const auto *rhsIntrinsic =
      static_cast<const GPUIntrinsicType *>(&rhs);
```

But this cast seems inherently unsafe -- if we need `GPUIntrinsicType`, we should change the function signature of `sortMMAIntrinsics` to take that instead.